rclcpp/include/rclcpp/node_impl.hpp

```diff
@@ -92,13 +95,40 @@ Node::create_subscription(
   const SubscriptionOptionsWithAllocator<AllocatorT> & options,
   typename MessageMemoryStrategyT::SharedPtr msg_mem_strat)
 {
-  return rclcpp::create_subscription<MessageT>(
-    *this,
-    extend_name_with_sub_namespace(topic_name, this->get_sub_namespace()),
-    qos,
-    std::forward<CallbackT>(callback),
-    options,
-    msg_mem_strat);
+  if (options.topic_stats_options.state == rclcpp::TopicStatisticsState::Enable) {
```

Comment:
I don't understand why this needs to be in the `rclcpp::Node` class. I know you said there were circular references or something preventing you from putting it into the `Subscription` class (I ran out of time investigating this, but I believe it should be possible with some rearranging or forward declaring), but this branching logic should at least be in `rclcpp::create_subscription` instead. Putting it here means that `rclcpp_lifecycle::LifecycleNode` does not get this feature, and it generally couples this feature to `rclcpp::Node` only, which is not ideal.

Reply:
I tried implementing in the `create_subscription` class as well and was getting compiler errors (similar issue when trying to create in the `Subscription` class). I'm again trying to see if it's possible in a branch, but simply moving the working code from `node_impl` to `create_subscription`

```
typename std::shared_ptr<SubscriptionT>
create_subscription(
  NodeT && node,
  const std::string & topic_name,
  const rclcpp::QoS & qos,
  CallbackT && callback,
  const rclcpp::SubscriptionOptionsWithAllocator<AllocatorT> & options = (
    rclcpp::SubscriptionOptionsWithAllocator<AllocatorT>()
  ),
  typename MessageMemoryStrategyT::SharedPtr msg_mem_strat = (
    MessageMemoryStrategyT::create_default()
  )
)
{
  using rclcpp::node_interfaces::get_node_topics_interface;
  auto node_topics = get_node_topics_interface(std::forward<NodeT>(node));

  std::shared_ptr<rclcpp::topic_statistics::SubscriptionTopicStatistics<CallbackMessageT>>
    subscription_topic_stats = nullptr;

  if (rclcpp::detail::resolve_enable_topic_statistics(options, *node.node_base_)) {
    auto  publisher =
      node.create_publisher<statistics_msgs::msg::MetricsMessage>(options.topic_stats_options.publish_topic, 10);

    auto subscription_topic_stats = std::make_shared<
      rclcpp::topic_statistics::SubscriptionTopicStatistics<CallbackMessageT>
    >(node.get_name(), publisher);

    auto timer = node.create_wall_timer(
      options.topic_stats_options.publish_period, [subscription_topic_stats]() {
        subscription_topic_stats->publish_message();
      });

    subscription_topic_stats->set_publisher_timer(timer);
  }
(rest of the method omitted)
```
 results in an error I can't get past yet

```
[ 12%] Building CXX object CMakeFiles/rclcpp.dir/src/rclcpp/executor.cpp.o
In file included from /home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/node_impl.hpp:40:0,
                 from /home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/node.hpp:1167,
                 from /home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/src/rclcpp/executor.cpp:25:
/home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/create_subscription.hpp: In function ‘std::shared_ptr<SubscriptionT> rclcpp::create_subscription(NodeT&&, const string&, const rclcpp::QoS&, CallbackT&&, const rclcpp::SubscriptionOptionsWithAllocator<AllocatorT>&, typename MessageMemoryStrategyT::SharedPtr)’:
/home/ubuntu/ros2_worksapce/src/ros2/rclcpp/rclcpp/include/rclcpp/create_subscription.hpp:80:65: error: expected primary-expression before ‘>’ token
       node.create_publisher<statistics_msgs::msg::MetricsMessage>(options.topic_stats_options.publish_topic, 10);
                                                                 ^
CMakeFiles/rclcpp.dir/build.make:403: recipe for target 'CMakeFiles/rclcpp.dir/src/rclcpp/executor.cpp.o' failed
make[2]: *** [CMakeFiles/rclcpp.dir/src/rclcpp/executor.cpp.o] Error 1
CMakeFiles/Makefile2:1182: recipe for target 'CMakeFiles/rclcpp.dir/all' failed
make[1]: *** [CMakeFiles/rclcpp.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

Reply:
Basically it comes down to the fact that we need a lot of things from the node level, create a publisher (seems we can use `rclcpp:create_publisher`?), the node name (ok, make a topic statistics parameter in the options), and also need to create a timer tied to the node. I don't see how that exists at the `create_subscription` level.

Reply:
You can definitely do it, but you will likely need to pass more interfaces into create_subscription, which makes total sense because the subscription is now pulling in all of these dependencies (creating a subscription now requires the interface to make timers because it now creates a timer...).

For the timer you can pass additionally the NodeTimerInterface and emulate this:

https://github.com/ros2/rclcpp/blob/649d72f835bfca334e7eab39a8c16752094d6317/rclcpp/include/rclcpp/node_impl.hpp#L104-L117

Ideally we would create a `rclcpp::create_*timer` function(s) to emulate how it works with publishers and subscriptions, and then use that. But in the meantime we'd have to copy what Node does above.

And for the publisher you can use `rclcpp::create_publisher`, as you suggested.

The node name can be accessed via the NodeBaseInterface, which can be accessed via the NodeTopicsInterface:

https://github.com/ros2/rclcpp/blob/649d72f835bfca334e7eab39a8c16752094d6317/rclcpp/include/rclcpp/node_interfaces/node_topics_interface.hpp#L79-L82

The other option would be to invert the dependencies, as I suggested, and instead register the Subscription with a statistics publisher which is created separately. This is close to what you have now, except instead of a separate stand alone class, you just have a publisher and a timer which are separately created for each subscription, so there's no chance for reuse between subscriptions.

Reply:
So, to summarize, you can either add the necessary interfaces to the `rclcpp::create_subscription()` function or leave it mostly as-is and update lifecycle node to also do what `rclcpp::Node::create_subscription()` is doing.

Reply:
@wjwwood thanks for your help: in the latest commit I was able to move everything over to `create_subscription` and out of `node_impl`.